samples/exit-handler-email/component.yaml

```diff
@@ -0,0 +1,58 @@
+name: sendmail
+description: |
+  This task sends a simple email to receivers via SMTP server
+inputs:
+  - {name: server, description: 'secret name for SMTP server information (url, port, password)'}
+  - {name: subject, description: 'plain text email subject'}
+  - {name: body, description: 'plain text email body'}
+  - {name: sender, description: 'sender email address'}
+  - {name: recipients, description: 'recipient email addresses (space delimited list)'}
+implementation:
+  container:
+    image: docker.io/library/python:3.8-alpine@sha256:c31682a549a3cc0a02f694a29aed07fd252ad05935a8560237aed99b8e87bf77 #tag: 3.8-alpine
+    command:
+    - python3
+    - -u
+    - -c
+    - |
+      #!/usr/bin/env python3
+      import smtplib, ssl, os
+      port = os.getenv('PORT')
+      smtp_server = os.getenv('SERVER')
+      sender_email = "$(params.sender)"
+      receiver_emails = "$(params.recipients)"
+      user = os.getenv('USER')
+      password = os.getenv('PASSWORD')
+      tls = os.getenv('TLS')
+      message = f"""\
+      Subject: $(params.subject)
+      To: {receiver_emails}
+      From: {sender_email}
+      $(params.body)"""
+      print(message)
+      if tls == 'True':
+          context = ssl.create_default_context()
+          server = smtplib.SMTP_SSL(smtp_server, port, context=context)
+      else:
+          server = smtplib.SMTP(smtp_server, port)
+      if password != '':
+          server.login(user, password)
+      for receiver in receiver_emails.split(' '):
+          server.sendmail(sender_email, receiver, message)
+      server.quit()
+
+      import argparse
+      _parser = argparse.ArgumentParser('sendmail inputs')
+      _parser.add_argument("--server", type=str, required=True)
+      _parser.add_argument("--subject", type=str, required=True)
+      _parser.add_argument("--body", type=str, required=True)
+      _parser.add_argument("--sender", type=str, required=True)
+      _parser.add_argument("--recipients", type=str, required=True)
```

Comment:
Thanks Tommy, could you please add attachment support?

Reply:
how do you want to pass the attachment in kfp? using big data passing (via workspace) or volumeOP? By default, big data passing and volumeOP will be create a new pvc to pass the attachment for each pipeline.

Reply:
big data passing is much more meanful for this case, as the case of the attachment should be from client, but not server.

Reply:
So right now here is the component.yaml for sending just one attachment file:
```yaml
name: sendmail
description: |
  This task sends a simple email to receivers via SMTP server
inputs:
  - {name: server, description: 'secret name for SMTP server information (url, port, password)'}
  - {name: subject, description: 'plain text email subject'}
  - {name: body, description: 'plain text email body'}
  - {name: sender, description: 'sender email address'}
  - {name: recipients, description: 'recipient email addresses (space delimited list)'}
  - {name: attachment_path, description: 'email attachment file path'}
implementation:
  container:
    image: docker.io/library/python:3.8-alpine@sha256:c31682a549a3cc0a02f694a29aed07fd252ad05935a8560237aed99b8e87bf77 #tag: 3.8-alpine
    command:
    - python3
    - -u
    - -c
    - |
      #!/usr/bin/env python3
      import argparse
      _parser = argparse.ArgumentParser('sendmail inputs')
      _parser.add_argument("--server", type=str, required=True)
      _parser.add_argument("--subject", type=str, required=True)
      _parser.add_argument("--body", type=str, required=True)
      _parser.add_argument("--sender", type=str, required=True)
      _parser.add_argument("--recipients", type=str, required=True)
      _parser.add_argument("--attachment_path", type=str, default='')
      _parsed_args = _parser.parse_args()

      import smtplib, ssl, os
      from pathlib import Path
      from email.mime.multipart import MIMEMultipart
      from email.mime.base import MIMEBase
      from email.mime.text import MIMEText
      from email.utils import COMMASPACE, formatdate
      from email import encoders
      port = os.getenv('PORT')
      smtp_server = os.getenv('SERVER')
      sender_email = "$(params.sender)"
      receiver_emails = "$(params.recipients)"
      user = os.getenv('USER')
      password = os.getenv('PASSWORD')
      tls = os.getenv('TLS')
      path = _parsed_args.attachment_path

      msg = MIMEMultipart()
      msg['From'] = sender_email
      msg['To'] = receiver_emails
      msg['Subject'] = "$(params.subject)"
      msg.attach(MIMEText("$(params.body)"))
      if tls == 'True':
          context = ssl.create_default_context()
          server = smtplib.SMTP_SSL(smtp_server, port, context=context)
      else:
          server = smtplib.SMTP(smtp_server, port)
      if password != '':
          server.login(user, password)
      part = MIMEBase('application', "octet-stream")
      with open(path, 'rb') as file:
          part.set_payload(file.read())
      encoders.encode_base64(part)
      part.add_header('Content-Disposition',
                      'attachment; filename="{}"'.format(Path(path).name))
      msg.attach(part)
      for receiver in receiver_emails.split(' '):
          server.sendmail(sender_email, receiver, msg.as_string())
      server.quit()
    args: [
      --server, {inputValue: server},
      --subject, {inputValue: subject},
      --body, {inputValue: body},
      --sender, {inputValue: sender},
      --recipients, {inputValue: recipients},
      --attachment_path, {inputPath: attachment_path},
    ]
```

Reply:
But if I want to make the containerOp or component.yaml to take one or more attachments from the previous output path, we will start seeing issue like https://github.com/kubeflow/kfp-tekton/issues/315 because kfp needs to have explicit arguments for their functions.

Reply:
A workaround I could think of is wrapping the python code into a custom container that can take a list arguments. Then, the user can define a new ContainerOp using that custom container to change the number of attachments in the list. This is less friendly than component.yaml, but at least we don't have to change the DSL code to support this feature.

Reply:
By the way, what service do you use for sending emails through SMTP? I have a trial version from Sendgrid, but having some weird problem when sending emails with attachments.

Reply:
One attachment should be OK for the example.

Reply:
so because we are running the notification task as exit handler, we cannot make sure any task will complete in the pipeline. Therefore, KFP doesn't allow dependency to be made for the exit task. 

Since we cannot create a dependency of the output path from a task, we cannot use the big data passing to map the workspace value for us. So right now I just create a pvc and assume the regular task will complete and write the file to the assumed path. It also can be done with volumeOp, but volumeOp will create a new pvc for every pipeline.

Reply:
Oh, yes, you are right, the exit handler has limitition here to run a pipeline behind it.